9.0.0-preview.3 fails to build when package source (`--with-packages`) is read-only

<!--
Please keep in mind that the GitHub issue tracker is not intended as a general support forum, but for reporting **non-security** bugs and feature requests.

If you believe you have an issue that affects the SECURITY of the platform, please do NOT create an issue and instead email your issue details to secure@microsoft.com. Your report may be eligible for our [bug bounty](https://www.microsoft.com/en-us/msrc/bounty-dot-net-core) but ONLY if it is reported through email.

For discussion or question, consider creating a new [discussion](https://github.com/dotnet/source-build/discussions).
-->

### Describe the Bug

If I use build.sh with `--with-packages [path]` and point it to a directory on a read-only file-system, I get an error like:

eng/tools/init-build.proj(53,5): error MSB3677: Unable to move file "[path]/SourceBuildReferencePackages/Humanizer.Core.2.2.0.nupkg" to "prereqs/packages/reference/Humanizer.Core.2.2.0.nupkg". Moving target is read-only

This hasn't been a problem on previous versions, including: 8.0.4, 9.0.0-preview.2.

--- a/src/SourceBuild/content/eng/init-source-only.proj
+++ b/src/SourceBuild/content/eng/init-source-only.proj
@@ -13,47 +13,63 @@
   <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
   <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
 
+  <PropertyGroup>
+    <ExternalTarballsDir>$([MSBuild]::NormalizeDirectory('$(PrereqsPackagesDir)', 'archive'))</ExternalTarballsDir>
+  </PropertyGroup>
+
   <Target Name="Build"
           DependsOnTargets="
       UnpackTarballs;
       BuildMSBuildSdkResolver;
       ExtractToolsetPackages" />
 
-  <Target Name="UnpackTarballs"
-          Inputs="$(MSBuildProjectFullPath)"
-          Outputs="$(BaseIntermediateOutputPath)UnpackTarballs.complete" >
-    <PropertyGroup>
-      <ExternalTarballsDir>$([MSBuild]::NormalizeDirectory('$(PrereqsPackagesDir)', 'archive'))</ExternalTarballsDir>
-    </PropertyGroup>
-
+  <!-- Unpacks the Private.SourceBuilt.Artifacts archive -->
+  <Target Name="UnpackSourceBuiltArtifactsArchive"
+          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''"
+          Inputs="$(ExternalTarballsDir)"
+          Outputs="$(PrebuiltSourceBuiltPackagesPath)">
     <MakeDir Directories="$(PrebuiltSourceBuiltPackagesPath)" Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''" />
     <Exec Command="tar -xzf $(ExternalTarballsDir)$(SourceBuiltArtifactsTarballName).*$(ArchiveExtension)"
-          WorkingDirectory="$(PrebuiltSourceBuiltPackagesPath)"
-          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''" />
+          WorkingDirectory="$(PrebuiltSourceBuiltPackagesPath)" />
+  </Target>
 
-    <!--
-      Check for a prebuilt dependency tarball and extract if exists. If there isn't one, we expect
-      the build to be working without prebuilts.
-    -->
+  <!-- Check for a prebuilt dependency tarball and extract if exists. If there isn't one, we expect
+       the build to be working without prebuilts. -->
+  <Target Name="UnpackSourceBuiltPrebuiltsArchive"
+          Inputs="$(ExternalTarballsDir)"
+          Outputs="$(PrebuiltPackagesPath)">
     <ItemGroup>
       <SourceBuiltPrebuiltsTarballFile Include="$(ExternalTarballsDir)$(SourceBuiltPrebuiltsTarballName).*$(ArchiveExtension)" />
     </ItemGroup>
+
     <Exec Command="tar -xzf %(SourceBuiltPrebuiltsTarballFile.FullPath)"
           WorkingDirectory="$(PrebuiltPackagesPath)"
           Condition="'@(SourceBuiltPrebuiltsTarballFile)' != ''" />
+  </Target>
 
-    <!-- Copy SBRP packages to reference packages location -->
+  <!-- Copy SBRP packages to reference packages location. -->
+  <Target Name="CopySourceBuiltReferencePackages"
+          DependsOnTargets="UnpackSourceBuiltArtifactsArchive"
+          Inputs="$(PrebuiltSourceBuiltPackagesPath)SourceBuildReferencePackages"
+          Outputs="$(ReferencePackagesDir)">
     <ItemGroup>
-      <UnpackedSourceBuildReferencePackages Include="$(PrebuiltSourceBuiltPackagesPath)SourceBuildReferencePackages/*"/>
+      <UnpackedSourceBuildReferencePackage Include="$(PrebuiltSourceBuiltPackagesPath)SourceBuildReferencePackages/*" />
     </ItemGroup>
 
-    <Move SourceFiles="@(UnpackedSourceBuildReferencePackages)" DestinationFiles="$(ReferencePackagesDir)%(Filename)%(Extension)" />
-
-    <MakeDir Directories="$(BaseIntermediateOutputPath)" />
-    <Touch Files="$(BaseIntermediateOutputPath)UnpackTarballs.complete" AlwaysCreate="true">
-      <Output TaskParameter="TouchedFiles" ItemName="FileWrites" />
-    </Touch>
+    <!-- Either move or copy the unpacked SBRP packages.
+         Don't move when the packages directory is externally provided. -->
+    <Move SourceFiles="@(UnpackedSourceBuildReferencePackage)"
+          DestinationFolder="$(ReferencePackagesDir)"
+          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''" />
+    <Copy SourceFiles="@(UnpackedSourceBuildReferencePackage)"
+          DestinationFolder="$(ReferencePackagesDir)"
+          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' != ''" />
   </Target>
+
+  <Target Name="UnpackTarballs"
+          DependsOnTargets="UnpackSourceBuiltArtifactsArchive;
+                            UnpackSourceBuiltPrebuiltsArchive;
+                            CopySourceBuiltReferencePackages" />
 
   <!-- Build the custom msbuild sdk resolver. -->
   <Target Name="BuildMSBuildSdkResolver"
@@ -74,10 +90,7 @@
   </Target>
 
   <!-- Extract toolset packages into the bootstrap folder -->
-  <Target Name="ExtractToolsetPackages"
-          DependsOnTargets="UnpackTarballs"
-          Inputs="$(MSBuildProjectFullPath)"
-          Outputs="$(BaseIntermediateOutputPath)ExtractToolsetPackages.complete">
+  <Target Name="ExtractToolsetPackages" DependsOnTargets="UnpackTarballs">
     <ItemGroup>
       <ToolsetPackage Include="Microsoft.DotNet.Arcade.Sdk" SourceFolder="$(PrebuiltSourceBuiltPackagesPath)" Version="$(ARCADE_BOOTSTRAP_VERSION)" />
       <ToolsetPackage Include="Microsoft.Build.NoTargets" SourceFolder="$(ReferencePackagesDir)" Version="$(NOTARGETS_BOOTSTRAP_VERSION)" />
@@ -87,11 +100,6 @@
     <Unzip SourceFiles="%(ToolsetPackage.SourceFolder)%(ToolsetPackage.Identity).%(ToolsetPackage.Version).nupkg"
            DestinationFolder="$(BootstrapPackagesDir)$([System.String]::Copy('%(ToolsetPackage.Identity)').ToLowerInvariant())/%(ToolsetPackage.Version)"
            SkipUnchangedFiles="true" />
-
-    <MakeDir Directories="$(BaseIntermediateOutputPath)" />
-    <Touch Files="$(BaseIntermediateOutputPath)ExtractToolsetPackages.complete" AlwaysCreate="true">
-      <Output TaskParameter="TouchedFiles" ItemName="FileWrites" />
-    </Touch>
   </Target>
 
 </Project>

--- a/src/SourceBuild/content/eng/init-source-only.proj
+++ b/src/SourceBuild/content/eng/init-source-only.proj
@@ -28,7 +28,7 @@
           Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''"
           Inputs="$(ExternalTarballsDir)"
           Outputs="$(PrebuiltSourceBuiltPackagesPath)">
-    <MakeDir Directories="$(PrebuiltSourceBuiltPackagesPath)" Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''" />
+    <MakeDir Directories="$(PrebuiltSourceBuiltPackagesPath)" />
     <Exec Command="tar -xzf $(ExternalTarballsDir)$(SourceBuiltArtifactsTarballName).*$(ArchiveExtension)"
           WorkingDirectory="$(PrebuiltSourceBuiltPackagesPath)" />
   </Target>

